File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque
import sys
read = sys.stdin.readline

N, K = map(int, read().rstrip().split())
maxNum = 100001
visited = [0] * maxNum

def findWay(x, way):
  X = [1, -1, 2*x, -2*x]
  if way >= 2:
    return X[way]
  return x + X[way]

def bfs(startX):
  queue = deque([[startX, 0]])
  visited[startX] = 1
  answer = 0

  while queue:
    x, cnt = queue.popleft()

    if x == K:
      answer = cnt
      break

    for i in range(4):
      nextX = findWay(x, i)
      if nextX < 0 or nextX >= maxNum: continue
      if visited[nextX]: continue
      visited[nextX] = 1
      queue.append([nextX, cnt + 1])

  return answer

print(bfs(N))

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
const fs = require('fs');
const [N, K] = fs.readFileSync('/dev/stdin').toString().trim().split(' ').map(Number);
const maxlen = 100000;
const minlen = 0;
const visited = Array(maxlen + 1).fill(0);

function bfs() {
  const queue = [];
  queue.push([N, 0]); // [N, depth]
  visited[N] = 1;
  
  while (queue.length) {
    const [n, depth] = queue.shift();

    if (n === K) return depth;

    const check = [n - 1, n + 1, n * 2];
    
    for (let i = 0; i < 3; i++) {
      if (check[i] >= minlen && check[i] <= maxlen) {
        if (!visited[check[i]]) {
          if (check[i] === K) return depth + 1;
          visited[check[i]] = 1;
          queue.push([check[i], depth + 1]);
        }
      }
    }
  }
}

console.log(bfs());